pkg/receive/handler.go

```diff
@@ -180,8 +191,82 @@ func (h *Handler) receive(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	if err := h.receiver.Receive(&wreq); err != nil {
+	tenant := r.Header.Get(tenantHeader)
+	local, err := h.forward(r.Context(), tenant, &wreq)
+	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
+	// There may be no WriteRequest destined for the local node.
+	if local != nil {
+		if err := h.receiver.Receive(local); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+// forward accepts a write request, batches its time series by
+// corresponding endpoint, and forwards them in parallel. It returns a write
+// request containing only the time series that correspond to
+// local handler.
+func (h *Handler) forward(ctx context.Context, tenant string, wreq *prompb.WriteRequest) (*prompb.WriteRequest, error) {
+	wreqs := make(map[string]*prompb.WriteRequest)
+	for i := range wreq.Timeseries {
+		endpoint, err := h.hashring.Get(tenant, &wreq.Timeseries[i])
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := wreqs[endpoint]; !ok {
+			wreqs[endpoint] = &prompb.WriteRequest{}
+		}
+		wr := wreqs[endpoint]
+		wr.Timeseries = append(wr.Timeseries, wreq.Timeseries[i])
+	}
+
+	ec := make(chan error)
+	defer close(ec)
+	var n int
```

Comment:
I am pretty unsure, what `n` in this case does. Would be nice to either document or giving it a more descriptive name.

Reply:
`n` like `i` is pretty standard for a count. I can change it to `i` instead if that’s clearer

Reply:
Yes, I got what it is used for, just some more guidance on why we need to have it next to the loop below, would be nice. (Similar to what you wrote to me on Slack, we don't want to use a WaitGroup because... This improves error handling compared to WaitGroups...) Something like that.

Reply:
done :+1: